src/jwe/mod.rs

```diff
@@ -0,0 +1,201 @@
+
+            let plaintext_ = msg
+                .decrypt::<CE, KDF, KE, KW>(Some((alice_kid, &alice_pkey)), (bob_kid, &bob_key))
+                .expect("unable decrypt.");
+
+            assert_eq!(plaintext_, plaintext.as_bytes());
+        }
+    }
+
+    #[test]
+    fn anoncrypt_works() {
+        _anoncrypt_works::<
+            AesKey<A128CbcHs256>,
+            EcdhEs<'_, X25519KeyPair>,
+            X25519KeyPair,
+            AesKey<A256Kw>,
+        >();
+
+        _anoncrypt_works::<
+            AesKey<A128CbcHs256>,
+            EcdhEs<'_, P256KeyPair>,
+            P256KeyPair,
+            AesKey<A256Kw>,
+        >();
+
+        _anoncrypt_works::<AesKey<A256Gcm>, EcdhEs<'_, X25519KeyPair>, X25519KeyPair, AesKey<A256Kw>>(
+        );
+
+        _anoncrypt_works::<AesKey<A256Gcm>, EcdhEs<'_, P256KeyPair>, P256KeyPair, AesKey<A256Kw>>();
+
+        _anoncrypt_works::<
+            Chacha20Key<XC20P>,
+            EcdhEs<'_, X25519KeyPair>,
+            X25519KeyPair,
+            AesKey<A256Kw>,
+        >();
+
+        _anoncrypt_works::<Chacha20Key<XC20P>, EcdhEs<'_, P256KeyPair>, P256KeyPair, AesKey<A256Kw>>(
+        );
+
+        /// TODO: P-384 and P-521 support after solving https://github.com/hyperledger/aries-askar/issues/10
+
+        fn _anoncrypt_works<CE, KDF, KE, KW>()
+        where
+            CE: KeyAeadInPlace + KeyAeadMeta + KeyGen + ToSecretBytes + KeySecretBytes,
+            KDF: JoseKDF<KE, KW>,
+            KE: KeyExchange + KeyGen + ToJwk + FromJwk + ToPublicBytes + KeyPublicBytes,
+            KW: KeyWrap + FromKeyDerivation,
+        {
+            let bob_kid = "did:example:bob#key-1";
+            let bob_key = KE::random().expect("unable random.");
+
+            let bob_pkey = {
+                let bytes = bob_key.to_public_bytes().expect("unable to_public_bytes.");
+
+                KE::from_public_bytes(&bytes).expect("unable from_public_bytes.")
+            };
+
+            let plaintext = "Some plaintext.";
+
+            let msg = jwe::encrypt::<CE, KDF, KE, KW>(
+                plaintext.as_bytes(),
+                Algorithm::Ecdh1esA256kw,
+                EncAlgorithm::A256cbcHs512,
+                None,
+                &[(bob_kid, &bob_pkey)],
+            )
+            .expect("unable encrypt.");
+
+            let mut buf = Vec::new();
+            let msg = jwe::parse(&msg, &mut buf).expect("unable parse.");
+
+            assert_eq!(msg.jwe.recipients.len(), 1);
+            assert_eq!(msg.jwe.recipients[0].header.kid, "did:example:bob#key-1");
+
+            assert_eq!(msg.protected.alg, Algorithm::Ecdh1esA256kw);
+            assert_eq!(msg.protected.enc, EncAlgorithm::A256cbcHs512);
+
+            assert_eq!(msg.skid, None);
+
+            let _plaintext = msg
+                .decrypt::<CE, KDF, KE, KW>(None, (bob_kid, &bob_key))
+                .expect("unable decrypt.");
+
+            assert_eq!(_plaintext, plaintext.as_bytes());
+        }
+    }
```

Comment:
It would be great to add the following tests
- Positive test encryption: multiple recipients
- Negative test encryption: sender and recipient have different key types (curves)
- Negative test encryption: multiple recipients with different key types (curves)
- Negative test encryption: non-A256CBC is passed for ECDH1-PU
- Negative test decryption: sender and recipient have different key types (curves)
- Negative test decryption: ciphertext is invalid
- Negative test decryption: passed recipient kid is not found in the per-recipient headers
- Negative test decryption: invalid message (no epk, no recipients, etc.)
- Negative test decryption: encrypted_key value for the the given recipient key is invalid 
- test decryption: encrypted_key value for another recipient (Charlie) is invalid while it's valid for the passed recipient (Bob). No exception expected.
- test decryption: unexpected values in JSON are met (aad in particular). No exception is expected (just ignore them?)
- Test decryption: multi-recipient encryption for two keys is done, but decryption is done for one key only (separately for Bob and for Charlie)
- test decryption: skid is not present, but (valid) apu is
- negative test decryption: neither skid nor apu are present
- negative test decryption: apu is invalid
- negative test decryption: apv is invalid

Reply:
- ~~Positive test encryption: multiple recipients~~
- ~~Negative test encryption: sender and recipient have different key types (curves)~~
  - ~~It is impossible case for jose helpers due typing~~
- Negative test encryption: multiple recipients with different key types (curves)
- ~~Negative test encryption: non-A256CBC is passed for ECDH1-PU~~
  - ~~It is out of scope for Jose helpers. Will be handled on pack/unpack level~~
- ~~Negative test decryption: sender and recipient have different key types (curves)~~
  - ~~It is impossible case for jose helpers due typing~~
- ~~Negative test decryption: ciphertext is invalid~~
- ~~Negative test decryption: passed recipient kid is not found in the per-recipient headers~~
- ~~Negative test decryption: invalid message (no epk, no recipients, etc.)~~
- ~~Negative test decryption: encrypted_key value for the the given recipient key is invalid~~
- test decryption: encrypted_key value for another recipient (Charlie) is invalid while it's valid for the passed recipient (Bob). No exception expected.
- test decryption: unexpected values in JSON are met (aad in particular). No exception is expected (just ignore them?)
- ~~Test decryption: multi-recipient encryption for two keys is done, but decryption is done for one key only (separately for Bob and for Charlie)~~
   - ~~Out of scope for helpers~~
- ~~test decryption: skid is not present, but (valid) apu is~~
- ~~negative test decryption: neither skid nor apu are present~~
- ~~negative test decryption: apu is invalid~~
- ~~negative test decryption: apv is invalid~~